### PR TITLE
[MDB Ignore] faction refactor.

### PR DESCRIPTION
**Pull request by Ghommie**
## About The Pull Request
Factions have been refactored into status traits. I have gutted out a couple useless factions like "Lizard", "vines" and "sewer" and define types for what previously were objects with mapedited factions in the process.

`faction_check_mob` has been removed and `faction_check` brought to `/datum`, since status traits are supported there. For mobs with mind datums, the factions of their mind datum will also be taken into account. This way I can put job factions in line with mob factions and remove some factions code from certain antag datums and outfits.
Some snowflake procs such as `porta_turret/in_faction()` and `/isvineimmune()` have been killed.

There're currently a handful of features that set factions to something else, so I decided to rename the `faction` list variable for mobs to `innate_factions` (also moved it to `/mob/living`). It's now used for faction traits with `INNATE_TRAIT` as source to allow the intrinsic factions of a living mob to be changed without screwing up extraneous sources such as clothing, items, jobs.

Mob ref factions have been made into a component. Unlike the old `src.faction = mob.faction` it doesn't permanently overwrite any faction. It uses a trait macro that works similarly to the previous method, but instead of being a one time thing, it'll call `faction_check` for the datum `parent` is bound to, so `parent` also benefits or suffers from its loss or gain of faction traits. In addition to that, there's a trait that can block the innate faction traits from `parent`. By default, only the trait macro and faction check are done.

I have tested it locally. No runtimes, Legions and its skulls were ignoring my character when one of the two had a matching faction trait or the component, but I think this should receive a testmerge first.

TODO:
- [x] Add documentation.
- [x] Add a unit test.
- ~~Possibly make this simpler and easier to manage.~~
- ~~Make sure it isn't too heavy on live servers.~~

## Why It's Good For The Game
Like pre-refactor weather immunities, faction suffers from being lists of elements. It doesn't keep track of where factions come from, and people use the OR operator instead of +, which actually makes things worse the moment the user has multiple sources of a faction (like what happens when the chaplain has both the carp nullrod and a carp suit). To add injury, unlike weather immunities factions can get reset all over the code. This is usually the case of spawners and mobs that spawn minions, where it's hardly a problem, but other features such lazarus-revived and the dominate spell also do that. It's a deluge of potential problems.

## Changelog
I don't even know if a CL should be added. I just know this is a jabberwock.

### CODE REVIEW

**Comment by spookydonut** [comment]
Still has todos

**Comment by Ghommie** [comment]
> 
> 
> Still has todos

File: code/modules/unit_tests/factions.dm [hunk 1]
```diff
@@ -0,0 +1,33 @@
+/datum/unit_test/factions
```
Comment (Mothblocks) [inline code/modules/unit_tests/factions.dm:1]: Include documentation here with what the test does.

**Review by Mothblocks — CHANGES_REQUESTED** [review]
Test looks good (though I don't know how I feel about allocating raw livings) aside from one thing.

**Review by Maurukas — APPROVED** [review]
Sorry for the conflicts

**Review by Watermelon914 — CHANGES_REQUESTED** [review]

**Comment by Ghommie** [comment]
Closing this since I don't have the time to tend to it right now and it has to be redone in a way that doesn't use traits.

**Comment by Ghommie** [comment]
Closing this since I don't have the time to tend to it right now and it has to be redone in a way that doesn't use traits.

**Pull request closed without merging**